### ISSUE TITLE
Improve JSDoc support for destructured parameters

### ISSUE THREAD

**Issue by shicks** [google/closure-compiler#1781]
Currently it is very cumbersome to use destructuring for function parameters with more than a couple properties, particular if many are optional:

```
@param {{a: (number|undefined), b: (number|undefined), c: (number|undefined)}} param1
```

Note that the `number=` syntax does not work, and in fact leads to #1762.

Ideally we could write
```
/**
@param {number=} a
@param {number=} b
@param {number=} c
*/
function({a = 1, b = 2, c = 3} = {}) {}
```

**Comment by Dominator008** [comment]
@shicks That looks ambiguous, I guess? What's the difference between
```js
/**
@param {number=} a
@param {number=} b
@param {number=} c
*/
function({a = 1, b = 2, c = 3} = {}) {}
```
and
```js
/**
@param {number=} a
@param {number=} b
@param {number=} c
*/
function(a, b, c) {}
```
?

**Comment by MatrixFrog** [comment]
I agree. I'd like to see the `=` be supported someday, but I don't mind the verbosity other than that because it keeps things clear IMO.

**Comment by WhyEvenTry** [comment]
What IS the support for destructured parameters anyway? In WebStorm for example I cannot use the first syntax given in the ticket at the top. It only works when I pretend there is no destructuring as if everything in the object is a normal parameter. I have yet to see a word about destructuring in http://usejsdoc.org/index.html

**Comment by MatrixFrog** [comment]
The `@param {{a: (number|undefined), b: (number|undefined), c: (number|undefined)}} param1` syntax is supported. Note that what Closure Compiler supports is pretty different from usejsdoc.org in general. From our official docs:

> The Closure Compiler's type language derives from the annotations used by the JSDoc document-generation tool, though it has since diverged. It now includes several annotations that JSDoc does not support, and vice versa.

https://developers.google.com/closure/compiler/docs/js-for-compiler

**Comment by WhyEvenTry** [comment]
What is `param1` in the above example? Because with destructuring there is nothing in that place.

I know that JSDoc syntax != Closure compiler syntax. Since WebStorm supports both I can file a ticket with them if there is a problem with support for either, and I use the union of the syntaxes of both.

**Comment by MatrixFrog** [comment]
Our "JSDoc" parser requires every param to have a name, so for destructuring params, we allow any name that's a valid JS identifier, even though that name won't appear in the (non-comment) JS code.

**Comment by peteblois** [comment]
I agree with the original syntax request- converting to destructured parameters is a common refactoring operation when the number of optional parameters grows but the loss of per-parameter documentation is a clear regression.

**Comment by teppeis** [comment]
FYI [JSDoc 3](http://usejsdoc.org/tags-param.html) provides following syntax:

```js
/**
 * Assign the project to an employee.
 * @param {Object} employee - The employee who is responsible for the project.
 * @param {string} employee.name - The name of the employee.
 * @param {string} employee.department - The employee's department.
 */
Project.prototype.assign = function({ name, department }) {
    // ...
};
```
ref: https://github.com/google/closure-compiler/issues/2584
